pkg/controllers/applicationfailover/rb_application_failover_controller.go

```diff
@@ -0,0 +1,259 @@
+
+// RBApplicationFailoverController is to sync ResourceBinding's application failover behavior.
+type RBApplicationFailoverController struct {
+	client.Client
+	EventRecorder      record.EventRecorder
+	RateLimiterOptions ratelimiterflag.Options
+
+	// workloadUnhealthyMap records which clusters the specific resource is in an unhealthy state
+	workloadUnhealthyMap *workloadUnhealthyMap
+	ResourceInterpreter  resourceinterpreter.ResourceInterpreter
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *RBApplicationFailoverController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling ResourceBinding %s.", req.NamespacedName.String())
+
+	binding := &workv1alpha2.ResourceBinding{}
+	if err := c.Client.Get(ctx, req.NamespacedName, binding); err != nil {
+		if apierrors.IsNotFound(err) {
+			resource, err := helper.ConstructClusterWideKey(binding.Spec.Resource)
+			if err != nil {
+				return controllerruntime.Result{Requeue: true}, err
+			}
+			c.workloadUnhealthyMap.delete(resource)
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	if !c.bindingFilter(binding) {
+		resource, err := helper.ConstructClusterWideKey(binding.Spec.Resource)
+		if err != nil {
+			return controllerruntime.Result{Requeue: true}, err
+		}
+		c.workloadUnhealthyMap.delete(resource)
+		return controllerruntime.Result{}, nil
+	}
+
+	retryDuration, err := c.syncBinding(binding)
+	if err != nil {
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	if retryDuration > 0 {
+		klog.V(4).Infof("Retry to check health status of the workload after %v minutes.", retryDuration.Minutes())
+		return controllerruntime.Result{RequeueAfter: retryDuration}, nil
+	}
+	return controllerruntime.Result{}, nil
+}
+
+func (c *RBApplicationFailoverController) detectFailure(clusters []string, tolerationSeconds *int32, resource keys.ClusterWideKey) (int32, []string) {
+	var needEvictClusters []string
+	duration := int32(math.MaxInt32)
+
+	for _, cluster := range clusters {
+		if !c.workloadUnhealthyMap.hasWorkloadBeenUnhealthy(resource, cluster) {
+			c.workloadUnhealthyMap.setTimeStamp(resource, cluster)
+			if duration > *tolerationSeconds {
+				duration = *tolerationSeconds
+			}
+			continue
+		}
+		// When the workload in a cluster is in an unhealthy state for more than the tolerance time,
+		// and the cluster has not been in the GracefulEvictionTasks,
+		// the cluster will be added to the list that needs to be evicted.
+		unHealthyTimeStamp := c.workloadUnhealthyMap.getTimeStamp(resource, cluster)
+		if metav1.Now().After(unHealthyTimeStamp.Add(time.Duration(*tolerationSeconds) * time.Second)) {
```

Comment:
I think it's better to use a single `metav1.Now()` consistently instead of using the latest time every time.

Reply:
like:
``` go
now:= metav1.Now()
now.After()
now.Sub()

```

Reply:
Please @chaunceyjiang take another look.

Reply:
Generally LGTM. Only a question. https://github.com/karmada-io/karmada/pull/3461#discussion_r1185872544

Reply:
In order to implement the `PreConditions` we have to record the read state that probably need to extend the API, but the API is not ready yet. As I mentioned at https://github.com/karmada-io/karmada/pull/3461#discussion_r1186612642.